qutip/core/data/trace.pyx

```diff
@@ -39,6 +49,28 @@ cpdef double complex trace_dense(Dense matrix) nogil except *:
     return trace
 
 
+cpdef double complex trace_oper_ket_csr(CSR matrix) nogil except *:
+    cdef int N = <int>sqrt(matrix.shape[0])
+    _check_shape_oper_ket(N, matrix)
+    cdef size_t row, ptr
```

Comment:
We're no longer using `ptr` here.

Perhaps we can make both the sparse and dense versions a bit clearer and simpler by doing:
```
for ptr in range(0, N * N, N + 1):
    ... 
```
I'm also happy with using stride for the sparse version too.

Reply:
Can't do range with step without the gil...
Cython can't detect the step sign at compilation, I tried using uint, but no luck...

Reply:
If by explicitly set stride you mean defining a variable for it instead of using `N+1` everywhere, yes I did the change.